cunumeric/deferred.py

```diff
@@ -503,6 +511,32 @@ def _create_indexing_array(
                         f"dimension {i} doesn't match to the shape of the"
                         f"index array which is {rhs.shape[i]}"
                     )
+
+            # if key or rhs are empty, return an empty array with correct shape
+            if key.size == 0 or rhs.size == 0:
+                if rhs.size == 0 and key.size != 0:
+                    # we need to calculate shape of the 0 dim of output region
+                    # even though the size of it is 0
+                    # this can potentially be replaced with COUNT_NONZERO
+                    s = key.nonzero()[0].size
+                else:
+                    s = 0
+
+                out_shape = (s,) + tuple(
+                    rhs.shape[i] for i in range(key.ndim, rhs.ndim)
+                )
+                out = cast(
+                    DeferredArray,
+                    self.runtime.create_empty_thunk(
+                        out_shape,
+                        rhs.dtype,
+                        inputs=[rhs],
+                    ),
+                )
+
+                out._fill_with_zero()
```

Comment:
since out is uninitialized, we need to `touch it` with write permissions even if its size is zero. I use `FILL` task for this

Reply:
Is it sufficient to do something like this?

```
out.fill(np.zeros((), dtype=out.dtype))
```

Then we don't need to add an extra argument to the FILL task.

Reply:
Yes, you are right. This is so simple! I have spent quite some time thinking how to best implement this and missed the fact that we can pass numpy arrays to deferred methods